### PR TITLE
[proposal] Implementing K.rnn(reverse_sequence=True)

### PR DESCRIPTION
**Pull request by FremyCompany**
Hi,

I am a Tensorflow.js user and I made a proposed change to enable RNNs to return their output sequence in the reversed order, which is especially useful if you use go_backwards.

> https://github.com/tensorflow/tfjs-layers/pull/517

It was mentioned there that compatibility with the Python version of Tensorflow was something they cared about, and therefore it was suggested to also create a pull request on the Python side to gather feedback, and see if this implementation is something your team would support.

I'm not a Python developer but could make the necessary changes I believe, though I'd rather get an approval before going deeper because this might take a fair chunk of my time since this change has to be propagated to the layers etc. Now if in addition to approving the idea someone else also wanted to help me out and advance this pull request, I'd appreciate it of course :)

### CODE REVIEW

**Comment by qlzh727** [comment]
Hi @FremyCompany, sorry for the late reply. 

The proposed API can easily be achieved by processing the output tensor with keras.backend.reverse(). I think the currently rnn() API is already having too much knob on it, and adding the "reverse_sequence" will further complicate it, especially with the combination of existing go_backwards.

**Comment by FremyCompany** [comment]
@qlzh727 Thanks for the reply. No problem with the delay.

I would agree with you, but doing that has significant performance implications however, while doing the work at the `rnn` level would be free of cost, because it happens before the `concat` which is why I was considering this.

**Comment by qlzh727** [comment]
I see, it does have some performance implication, but I am not sure how significant it is. Adding @fchollet for this since it also affect the keras-team/keras.

**Comment by FremyCompany** [comment]
The impact might possibly be larger in tf.js than on the other implementation because there is no built-in reverse layer, and custom functions might be slower there than there are in Python (and they don't serialize) but this could be fixed at the tf.js level by introducing a built-in reverse layer I guess. If this group believes this is a better avenue to fix this, that's probably fine as well, but it would be slower.

**Comment by qlzh727** [comment]
I checked with Francois about this, and we think the proposed API change will be quite confusing, and we prefer the readability over slight performance issue.

The issue here is because of the combination of "reverse_sequence" and "go_backwards". We can't easily set default reverse_sequence to False, since it will modify the existing behavior when go_backwards is True. We could set default to None, and then choose the value based on go_backwards. However, at that stage, the API is bit complicated, and require user to read detailed API docstring to understand the combinations of the params, whereas usually params should be self contained and not affecting each other. Ideally, user should be able to just read the method signature, and understand the behavior of each parameter.

Performance wise, we are already doing reverse input and mask when go_backwards is True, we don't foresee any big issue if user has to reverse the output sequence by themselves. If tf.js is implementing same API surface as keras does, it should have the coverage of keras.backend.reverse(), which is also just a simple wrapper around array_ops.reverse().

**Comment by FremyCompany** [comment]
Okay, thanks for investigating!
